File: app/src/main/java/bassamalim/hidaya/fragments/BookChaptersFragment.java
package bassamalim.hidaya.fragments;

import android.content.Intent;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.appcompat.widget.SearchView;
import androidx.fragment.app.Fragment;
import androidx.preference.PreferenceManager;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.google.gson.Gson;

import java.util.ArrayList;

import bassamalim.hidaya.activities.BookViewer;
import bassamalim.hidaya.adapters.BookChapterAdapter;
import bassamalim.hidaya.databinding.FragmentBookChaptersBinding;
import bassamalim.hidaya.enums.ListType;
import bassamalim.hidaya.models.Book;
import bassamalim.hidaya.models.BookChapterCard;
import bassamalim.hidaya.other.Utils;

public class BookChaptersFragment extends Fragment {

    private FragmentBookChaptersBinding binding;
    private RecyclerView recycler;
    private BookChapterAdapter adapter;
    private final ListType type;
    private final int bookId;
    private Book book;
    private boolean[] favs;

    public BookChaptersFragment(ListType type, int bookId) {
        this.type = type;
        this.bookId = bookId;
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {

        binding = FragmentBookChaptersBinding.inflate(inflater, container, false);

        setupRecycler();

        setSearchListeners();

        return binding.getRoot();
    }

    @Override
    public void setMenuVisibility(boolean menuVisible) {
        super.setMenuVisibility(menuVisible);

        if (menuVisible) {
            adapter = new BookChapterAdapter(getContext(), makeCards(), bookId);
            recycler.setAdapter(adapter);
        }
    }

    private void getData() {
        String path = requireContext().getExternalFilesDir(null) + "/Books/" +
                bookId  + ".json";
        String jsonStr = Utils.getJsonFromDownloads(path);
        Gson gson = new Gson();
        book = gson.fromJson(jsonStr, Book.class);

        SharedPreferences pref = PreferenceManager.getDefaultSharedPreferences(requireContext());
        String favsStr = pref.getString("book" + bookId + "_favs", "");
        if (favsStr.length() == 0)
            favs = new boolean[book.getChapters().length];
        else
            favs = gson.fromJson(favsStr, boolean[].class);
    }

    private ArrayList<BookChapterCard> makeCards() {
        getData();

        ArrayList<BookChapterCard> cards = new ArrayList<>();
        for (int i = 0; i < book.getChapters().length; i++) {
            if (type == ListType.All || (type == ListType.Favorite && favs[i])) {
                String chapterTitle = book.getChapters()[i].getChapterTitle();

                int finalI = i;
                View.OnClickListener listener = v -> {
                    Intent intent = new Intent(getContext(), BookViewer.class);
                    intent.putExtra("book_id", bookId);
                    intent.putExtra("book_title", chapterTitle);
                    intent.putExtra("chapter_id", finalI);
                    startActivity(intent);
                };

                cards.add(new BookChapterCard(i, chapterTitle, favs[i], listener));
            }
        }
        return cards;
    }

    private void setupRecycler() {
        recycler = binding.recycler;
        LinearLayoutManager layoutManager = new LinearLayoutManager(getContext());
        recycler.setLayoutManager(layoutManager);
        adapter = new BookChapterAdapter(getContext(), makeCards(), bookId);
        recycler.setAdapter(adapter);
    }

    private void setSearchListeners() {
        binding.searchView.setOnQueryTextListener(new SearchView.OnQueryTextListener() {
            @Override
            public boolean onQueryTextSubmit(String query) {
                adapter.filter(query);
                return true;
            }
            @Override
            public boolean onQueryTextChange(String newText) {
                adapter.filter(newText);
                return true;
            }
        });
    }

    public void onDestroy() {
        super.onDestroy();
        binding = null;
        recycler.setAdapter(null);
        adapter = null;
    }
}

File: app/src/main/java/bassamalim/hidaya/fragments/BookChaptersFragment.kt
package bassamalim.hidaya.fragments

import android.content.Intent
import android.content.SharedPreferences
import android.os.Bundle
import android.view.*
import androidx.appcompat.widget.SearchView
import androidx.fragment.app.Fragment
import androidx.preference.PreferenceManager
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import bassamalim.hidaya.activities.BookViewer
import bassamalim.hidaya.adapters.BookChapterAdapter
import bassamalim.hidaya.databinding.FragmentBookChaptersBinding
import bassamalim.hidaya.enums.ListType
import bassamalim.hidaya.models.Book
import bassamalim.hidaya.models.BookChapter
import bassamalim.hidaya.other.Utils
import com.google.gson.Gson

class BookChaptersFragment(private val type: ListType, private val bookId: Int) : Fragment() {

    private var binding: FragmentBookChaptersBinding? = null
    private var recycler: RecyclerView? = null
    private var adapter: BookChapterAdapter? = null
    private lateinit var book: Book
    private lateinit var favs: BooleanArray

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?
    ): View {
        binding = FragmentBookChaptersBinding.inflate(inflater, container, false)

        setupRecycler()

        setSearchListeners()

        return binding!!.root
    }

    override fun setMenuVisibility(menuVisible: Boolean) {
        super.setMenuVisibility(menuVisible)

        if (menuVisible) {
            adapter = BookChapterAdapter(requireContext(), makeCards(), bookId)
            recycler!!.adapter = adapter
        }
    }

    private fun getData() {
        val path = requireContext().getExternalFilesDir(null).toString() + "/Books/" +
                bookId + ".json"
        val jsonStr = Utils.getJsonFromDownloads(path)
        val gson = Gson()
        book = gson.fromJson(jsonStr, Book::class.java)

        val pref: SharedPreferences =
            PreferenceManager.getDefaultSharedPreferences(requireContext())
        val favsStr: String = pref.getString("book" + bookId + "_favs", "")!!
        favs =
            if (favsStr.isEmpty()) BooleanArray(book.chapters.size)
            else gson.fromJson(favsStr, BooleanArray::class.java)
    }

    private fun makeCards(): ArrayList<BookChapter> {
        getData()

        val cards = ArrayList<BookChapter>()
        for (i in book.chapters.indices) {
            if (type == ListType.All || type == ListType.Favorite && favs[i]) {
                val chapterTitle = book.chapters[i].chapterTitle

                val listener = View.OnClickListener {
                    val intent = Intent(context, BookViewer::class.java)
                    intent.putExtra("book_id", bookId)
                    intent.putExtra("book_title", chapterTitle)
                    intent.putExtra("chapter_id", i)
                    startActivity(intent)
                }

                cards.add(BookChapter(i, chapterTitle, favs[i], listener))
            }
        }
        return cards
    }

    private fun setupRecycler() {
        recycler = binding!!.recycler
        val layoutManager = LinearLayoutManager(context)
        recycler!!.layoutManager = layoutManager
        adapter = BookChapterAdapter(requireContext(), makeCards(), bookId)
        recycler!!.adapter = adapter
    }

    private fun setSearchListeners() {
        binding!!.searchView.setOnQueryTextListener(object : SearchView.OnQueryTextListener {
            override fun onQueryTextSubmit(query: String): Boolean {
                adapter?.filter(query)
                return true
            }

            override fun onQueryTextChange(newText: String): Boolean {
                adapter?.filter(newText)
                return true
            }
        })
    }

    override fun onDestroyView() {
        super.onDestroyView()
        binding = null
        recycler!!.adapter = null
        adapter = null
    }

}